File: airbyte-commons/src/main/java/io/airbyte/commons/yaml/Yamls.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.yaml;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.JsonToken;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SequenceWriter;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
import com.fasterxml.jackson.dataformat.yaml.YAMLParser;
import com.google.common.collect.AbstractIterator;
import io.airbyte.commons.jackson.MoreMappers;
import io.airbyte.commons.lang.CloseableConsumer;
import io.airbyte.commons.lang.Exceptions;
import io.airbyte.commons.util.AutoCloseableIterator;
import io.airbyte.commons.util.AutoCloseableIterators;
import java.io.IOException;
import java.io.InputStream;
import java.io.Writer;
import java.util.Iterator;

@SuppressWarnings("PMD.AvoidBranchingStatementAsLastInLoop")
public class Yamls {

  private static final YAMLFactory YAML_FACTORY = new YAMLFactory();
  private static final ObjectMapper OBJECT_MAPPER = MoreMappers.initYamlMapper(YAML_FACTORY);

  private static final YAMLFactory YAML_FACTORY_WITHOUT_QUOTES = new YAMLFactory().enable(YAMLGenerator.Feature.MINIMIZE_QUOTES);
  private static final ObjectMapper OBJECT_MAPPER_WITHOUT_QUOTES = MoreMappers.initYamlMapper(YAML_FACTORY_WITHOUT_QUOTES);

  /**
   * Serialize object to YAML string. String values WILL be wrapped in double quotes.
   *
   * @param object - object to serialize
   * @return YAML string version of object
   */
  public static <T> String serialize(final T object) {
    try {
      return OBJECT_MAPPER.writeValueAsString(object);
    } catch (final JsonProcessingException e) {
      throw new RuntimeException(e);
    }
  }

  /**
   * Serialize object to YAML string. String values will NOT be wrapped in double quotes.
   *
   * @param object - object to serialize
   * @return YAML string version of object
   */
  public static String serializeWithoutQuotes(final Object object) {
    try {
      return OBJECT_MAPPER_WITHOUT_QUOTES.writeValueAsString(object);
    } catch (final JsonProcessingException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> T deserialize(final String yamlString, final Class<T> klass) {
    try {
      return OBJECT_MAPPER.readValue(yamlString, klass);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> T deserialize(final String yamlString, final TypeReference<T> typeReference) {
    try {
      return OBJECT_MAPPER.readValue(yamlString, typeReference);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static JsonNode deserialize(final String yamlString) {
    try {
      return OBJECT_MAPPER.readTree(yamlString);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  public static AutoCloseableIterator<JsonNode> deserializeArray(final InputStream stream) {
    try {
      final YAMLParser parser = YAML_FACTORY.createParser(stream);

      // Check the first token
      if (parser.nextToken() != JsonToken.START_ARRAY) {
        throw new IllegalStateException("Expected content to be an array");
      }

      final Iterator<JsonNode> iterator = new AbstractIterator<>() {

        @Override
        protected JsonNode computeNext() {
          try {
            while (parser.nextToken() != JsonToken.END_ARRAY) {
              return parser.readValueAsTree();
            }
          } catch (final IOException e) {
            throw new RuntimeException(e);
          }
          return endOfData();
        }

      };

      return AutoCloseableIterators.fromIterator(iterator, parser::close, null);

    } catch (final IOException e) {
      throw new RuntimeException(e);
    }
  }

  // todo (cgardens) - share this with Jsons if ever needed.

  /**
   * Creates a consumer that writes list items to the writer in a streaming fashion.
   *
   * @param writer writer to write to
   * @param <T> type of items being written
   * @return consumer that is able to write element to a list element by element. must be closed!
   */
  public static <T> CloseableConsumer<T> listWriter(final Writer writer) {
    return new YamlConsumer<>(writer, OBJECT_MAPPER);
  }

  public static class YamlConsumer<T> implements CloseableConsumer<T> {

    private final SequenceWriter sequenceWriter;

    public YamlConsumer(final Writer writer, final ObjectMapper objectMapper) {
      this.sequenceWriter = Exceptions.toRuntime(() -> objectMapper.writer().writeValuesAsArray(writer));

    }

    @Override
    public void accept(final T t) {
      try {
        sequenceWriter.write(t);
      } catch (final IOException e) {
        throw new RuntimeException(e);
      }
    }

    @Override
    public void close() throws Exception {
      // closing the SequenceWriter closes the Writer that it wraps.
      sequenceWriter.close();
    }

  }

}

File: airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/yaml/Yamls.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.commons.yaml

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.core.JsonToken
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SequenceWriter
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator
import com.google.common.collect.AbstractIterator
import io.airbyte.commons.concurrency.VoidCallable
import io.airbyte.commons.jackson.MoreMappers.initYamlMapper
import io.airbyte.commons.lang.CloseableConsumer
import io.airbyte.commons.lang.Exceptions.toRuntime
import io.airbyte.commons.util.AutoCloseableIterator
import io.airbyte.commons.util.AutoCloseableIterators
import java.io.IOException
import java.io.InputStream
import java.io.Writer

object Yamls {
    private val YAML_FACTORY = YAMLFactory()
    private val OBJECT_MAPPER = initYamlMapper(YAML_FACTORY)

    private val YAML_FACTORY_WITHOUT_QUOTES: YAMLFactory =
        YAMLFactory().enable(YAMLGenerator.Feature.MINIMIZE_QUOTES)
    private val OBJECT_MAPPER_WITHOUT_QUOTES = initYamlMapper(YAML_FACTORY_WITHOUT_QUOTES)

    /**
     * Serialize object to YAML string. String values WILL be wrapped in double quotes.
     *
     * @param object
     * - object to serialize
     * @return YAML string version of object
     */
    fun <T> serialize(`object`: T): String {
        try {
            return OBJECT_MAPPER.writeValueAsString(`object`)
        } catch (e: JsonProcessingException) {
            throw RuntimeException(e)
        }
    }

    /**
     * Serialize object to YAML string. String values will NOT be wrapped in double quotes.
     *
     * @param object
     * - object to serialize
     * @return YAML string version of object
     */
    fun serializeWithoutQuotes(`object`: Any?): String {
        try {
            return OBJECT_MAPPER_WITHOUT_QUOTES.writeValueAsString(`object`)
        } catch (e: JsonProcessingException) {
            throw RuntimeException(e)
        }
    }

    fun <T> deserialize(yamlString: String?, klass: Class<T>?): T {
        try {
            return OBJECT_MAPPER.readValue(yamlString, klass)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    fun <T> deserialize(yamlString: String?, typeReference: TypeReference<T>?): T {
        try {
            return OBJECT_MAPPER.readValue(yamlString, typeReference)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    @JvmStatic
    fun deserialize(yamlString: String?): JsonNode {
        try {
            return OBJECT_MAPPER.readTree(yamlString)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    fun deserializeArray(stream: InputStream?): AutoCloseableIterator<JsonNode> {
        try {
            val parser = YAML_FACTORY.createParser(stream)

            // Check the first token
            check(parser.nextToken() == JsonToken.START_ARRAY) { "Expected content to be an array" }

            val iterator: Iterator<JsonNode> =
                object : AbstractIterator<JsonNode>() {
                    override fun computeNext(): JsonNode? {
                        try {
                            while (parser.nextToken() != JsonToken.END_ARRAY) {
                                return parser.readValueAsTree()
                            }
                        } catch (e: IOException) {
                            throw RuntimeException(e)
                        }
                        return endOfData()
                    }
                }

            return AutoCloseableIterators.fromIterator<JsonNode>(
                iterator,
                VoidCallable { parser.close() },
                null
            )
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    // todo (cgardens) - share this with Jsons if ever needed.
    /**
     * Creates a consumer that writes list items to the writer in a streaming fashion.
     *
     * @param writer writer to write to
     * @param <T> type of items being written
     * @return consumer that is able to write element to a list element by element. must be closed!
     * </T>
     */
    fun <T> listWriter(writer: Writer?): CloseableConsumer<T> {
        return YamlConsumer(writer, OBJECT_MAPPER)
    }

    class YamlConsumer<T>(writer: Writer?, objectMapper: ObjectMapper) : CloseableConsumer<T> {
        private val sequenceWriter: SequenceWriter =
            toRuntime(callable = { objectMapper.writer().writeValuesAsArray(writer) })

        override fun accept(t: T) {
            try {
                sequenceWriter.write(t)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }

        @Throws(Exception::class)
        override fun close() {
            // closing the SequenceWriter closes the Writer that it wraps.
            sequenceWriter.close()
        }
    }
}
